// catboost/build_scripts/out/build.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildLocal = exports.buildNative = exports.buildModel = exports.compileBindings = exports.compileJs = void 0;
var common_1 = require("./common");
var fs_1 = require("fs");
var download_1 = require("./download");
var config_1 = require("./config");
var path_1 = require("path");
function compileTypeScript() {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, common_1.execProcess('npm run tsc')];
                case 1:
                    result = _b.sent();
                    if (result.code !== 0) {
                        console.error("Building ts library failed:\n            " + result.code + " " + result.signal + " " + ((_a = result.err) === null || _a === void 0 ? void 0 : _a.message));
                        process.exit(1);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function copyBindings() {
    fs_1.writeFileSync('./lib/catboost.d.ts', fs_1.readFileSync('./bindings/catboost.d.ts'));
    fs_1.writeFileSync('./lib/catboost.js', fs_1.readFileSync('./bindings/catboost.js'));
}
function compileNativeAddon(srcPath) {
    var _a;
    if (srcPath === void 0) { srcPath = path_1.join('..', '..'); }
    return __awaiter(this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    process.env['CATBOOST_SRC_PATH'] = srcPath;
                    return [4 /*yield*/, common_1.execProcess('node-gyp build')];
                case 1:
                    result = _b.sent();
                    if (result.code !== 0) {
                        console.error("Building native-addon library failed:\n            " + result.code + " " + result.signal + " " + ((_a = result.err) === null || _a === void 0 ? void 0 : _a.message));
                        process.exit(1);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function compileJs() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, compileTypeScript()];
                case 1:
                    _a.sent();
                    copyBindings();
                    return [2 /*return*/];
            }
        });
    });
}
exports.compileJs = compileJs;
function compileBindings() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, compileJs()];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, compileNativeAddon()];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.compileBindings = compileBindings;
function buildModelInterfaceLibrary(buildNativeExtraArgs) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var _i, buildNativeExtraArgs_1, buildNativeArg, srcPath, result;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    for (_i = 0, buildNativeExtraArgs_1 = buildNativeExtraArgs; _i < buildNativeExtraArgs_1.length; _i++) {
                        buildNativeArg = buildNativeExtraArgs_1[_i];
                        if (buildNativeArg.match(/^\-\-(targets|build\-root\-dir)/)) {
                            console.error("build_native extra arguments cannot contain --targets or --build-root-dir, they are predefined");
                            process.exit(1);
                        }
                    }
                    srcPath = process.env['CATBOOST_SRC_PATH'] || path_1.join('..', '..');
                    return [4 /*yield*/, common_1.execProcess("python3 " + path_1.join(srcPath, 'build', 'build_native.py') + " --targets catboostmodel --build-root-dir ./build "
                            + buildNativeExtraArgs.map(function (arg) { return '"' + arg + '"'; }).join(' '))];
                case 1:
                    result = _b.sent();
                    if (result.code !== 0) {
                        console.error("Building catboostmodel library failed:\n            " + result.code + " " + result.signal + " " + ((_a = result.err) === null || _a === void 0 ? void 0 : _a.message));
                        process.exit(1);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function configureGyp(srcPath) {
    var _a;
    if (srcPath === void 0) { srcPath = path_1.join('..', '..'); }
    return __awaiter(this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    process.env['CATBOOST_SRC_PATH'] = srcPath;
                    return [4 /*yield*/, common_1.execProcess('node-gyp configure')];
                case 1:
                    result = _b.sent();
                    if (result.code !== 0) {
                        console.error("Node-gyp configuration failed:\n            " + result.code + " " + result.signal + " " + ((_a = result.err) === null || _a === void 0 ? void 0 : _a.message));
                        process.exit(1);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function buildModel(buildNativeExtraArgs) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, configureGyp()];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, buildModelInterfaceLibrary(buildNativeExtraArgs)];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.buildModel = buildModel;
/** Build binary from repository. */
function buildNative(buildNativeExtraArgs) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, configureGyp()];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, buildModelInterfaceLibrary(buildNativeExtraArgs)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, compileBindings()];
                case 3:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.buildNative = buildNative;
function preparePlatformBinary(platform, arch) {
    return __awaiter(this, void 0, void 0, function () {
        var config, platform_arch, _i, _a, binary;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    config = config_1.readConfig();
                    if (platform == 'darwin') {
                        arch = 'universal2';
                    }
                    platform_arch = platform + '-' + arch;
                    _i = 0, _a = config.binaries[platform_arch];
                    _b.label = 1;
                case 1:
                    if (!(_i < _a.length)) return [3 /*break*/, 4];
                    binary = _a[_i];
                    return [4 /*yield*/, download_1.downloadBinaryFile('./build/catboost/libs/model_interface', binary)];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4:
                    if (platform == 'linux') {
                        fs_1.linkSync('./build/catboost/libs/model_interface/libcatboostmodel.so', './build/catboost/libs/model_interface/libcatboostmodel.so.1');
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function buildLocal(platform, arch) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, preparePlatformBinary(platform, arch)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, configureGyp('./inc')];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, compileNativeAddon('./inc')];
                case 3:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.buildLocal = buildLocal;
